tests/trace/openai/test_instrumentor.py

```diff
@@ -0,0 +1,377 @@
+        },
+        status=200,
+    )
+
+    response = openai.ChatCompletion.create(
+        model=model, messages=messages, temperature=temperature, functions=functions
+    )
+
+    response_text = response.choices[0]["message"]["content"]
+    assert "22" in response_text and "Boston" in response_text
+
+    spans = list(tracer.get_spans())
+    assert len(spans) == 1
+    span = spans[0]
+    attributes = span.attributes
+
+    assert span.span_kind is SpanKind.LLM
+    assert span.status_code == SpanStatusCode.OK
+    assert attributes[LLM_INPUT_MESSAGES] == [
+        {MESSAGE_ROLE: "user", MESSAGE_CONTENT: "What is the weather like in Boston?"},
+        {
+            MESSAGE_ROLE: "assistant",
+            MESSAGE_CONTENT: None,
+            MESSAGE_FUNCTION_CALL_NAME: "get_current_weather",
+            MESSAGE_FUNCTION_CALL_ARGUMENTS_JSON: '{"location": "Boston, MA"}',
+        },
+        {
+            MESSAGE_ROLE: "function",
+            MESSAGE_NAME: "get_current_weather",
+            MESSAGE_CONTENT: '{"temperature": "22", "unit": "celsius", "description": "Sunny"}',
+        },
+    ]
+    assert attributes[LLM_INVOCATION_PARAMETERS] == {
+        "model": model,
+        "messages": messages,
+        "temperature": temperature,
+        "functions": functions,
+    }
+    assert LLM_FUNCTION_CALL not in attributes
+    assert span.events == []
+
+
+@responses.activate
+def test_openai_instrumentor_records_authentication_error(
+    reload_openai_api_requestor, openai_api_key
+) -> None:
+    tracer = Tracer()
+    OpenAIInstrumentor(tracer).instrument()
+    responses.post(
+        "https://api.openai.com/v1/chat/completions",
+        json={
+            "error": {
+                "message": "error-message",
+                "type": "invalid_request_error",
+                "param": None,
+                "code": "invalid_api_key",
+            }
+        },
+        status=401,
+    )
+    model = "gpt-4"
+    messages = [{"role": "user", "content": "Who won the World Cup in 2018?"}]
+    temperature = 0.23
+
+    with pytest.raises(AuthenticationError):
+        openai.ChatCompletion.create(model=model, messages=messages, temperature=temperature)
+
+    spans = list(tracer.get_spans())
+    assert len(spans) == 1
+    span = spans[0]
+
+    events = span.events
+    assert len(events) == 1
+    event = events[0]
+    assert isinstance(event, SpanException)
+    attributes = event.attributes
+    assert attributes[EXCEPTION_TYPE] == "AuthenticationError"
+    assert attributes[EXCEPTION_MESSAGE] == "error-message"
+    assert "Traceback" in attributes[EXCEPTION_STACKTRACE]
+
+
+@responses.activate
+def test_openai_instrumentor_instrument_method_is_idempotent(
+    reload_openai_api_requestor, openai_api_key
+) -> None:
+    tracer = Tracer()
+    OpenAIInstrumentor(tracer).instrument()
+    OpenAIInstrumentor(tracer).instrument()
```

Comment:
This line is duplicated

Reply:
This one is testing that the instrumentation step is idempotent.

Reply:
oh I see, thanks! maybe a comment then?

Reply:
addressed in 990f3a6a